src/runtime/float.cpp

```diff
@@ -1424,6 +1425,385 @@ extern "C" PyObject* _Py_double_round(double x, int ndigits) noexcept {
     return result;
 }
 
+/* Case-insensitive locale-independent string match used for nan and inf
+   detection. t should be lower-case and null-terminated.  Return a nonzero
+   result if the first strlen(t) characters of s match t and 0 otherwise. */
+
+static int case_insensitive_match(const char* s, const char* t) {
+    while (*t && Py_TOLOWER(*s) == *t) {
+        s++;
+        t++;
+    }
+    return *t ? 0 : 1;
+}
+
+static char char_from_hex(int x) {
+    assert(0 <= x && x < 16);
+    return "0123456789abcdef"[x];
+}
+
+static int hex_from_char(char c) {
+    int x;
+    switch (c) {
+        case '0':
+            x = 0;
+            break;
+        case '1':
+            x = 1;
+            break;
+        case '2':
+            x = 2;
+            break;
+        case '3':
+            x = 3;
+            break;
+        case '4':
+            x = 4;
+            break;
+        case '5':
+            x = 5;
+            break;
+        case '6':
+            x = 6;
+            break;
+        case '7':
+            x = 7;
+            break;
+        case '8':
+            x = 8;
+            break;
+        case '9':
+            x = 9;
+            break;
+        case 'a':
+        case 'A':
+            x = 10;
+            break;
+        case 'b':
+        case 'B':
+            x = 11;
+            break;
+        case 'c':
+        case 'C':
+            x = 12;
+            break;
+        case 'd':
+        case 'D':
+            x = 13;
+            break;
+        case 'e':
+        case 'E':
+            x = 14;
+            break;
+        case 'f':
+        case 'F':
+            x = 15;
+            break;
+        default:
+            x = -1;
+            break;
+    }
+    return x;
+}
+
+// From CPython floatobject.c
```

Comment:
hmm my recommendation is for cases like this, just copy the CPython code unmodified and don't adapt any of it to our APIs.  both float_fromhex and float_hex have some minor issues with exception handling (throw both C API and Pyston exceptions), and in general I think the less we diverge from CPython the better :)

in this particular case, you could add a separate floatFromHex() function that just calls CPython's from_fromhex and then only have a single place that we need to convert exceptions.  or probably even better, is copy from CPython all the way up to PyMethodDef, and let BoxedMethodDescriptor handle all of the details of converting between CPython and Pyston sides.  check out str.cpp::string_methods for an example :)